### DISCUSSION TITLE
Upgrade Keycloak from 15.1.1 to 19.02 on Kubernetes with AWS RDS(Postgres)

### DISCUSSION THREAD

**Discussion by sirishkumar** [keycloak/keycloak#14682] [Archive]
Hi,

We have a Keycloak 15.1.1 deployment on Kubernetes with multiple replicas with the AWS RDS Postgres 13 Database backend. I did not find any upgrade guide or experience of other people regarding this setup or even with Kubernetes with Postgres using PVC upgrading Keycloak with multiple major version changes. 

Does anyone have any experience with the Keycloak upgrade on Kubernetes?  

I went through the change log and was able to run Keycloak locally using docker-compose only in HTTP mode as we terminate SSL at the reverse proxy.   

From upgrade instructions from Keycloak [documentation](https://www.keycloak.org/docs/latest/upgrading/index.html) is the following strategy is the right one without losing any data

1. Update the docker image with a new image running only in HTTP mode in our helm charts
2. Initially start the new deployment with only a single replica so that the database schema changes are applied

     `kc.sh start --spi-connections-jpa-default-migration-strategy=update`

    > When I tried to upgrade my local deployment with the above command, Keycloak was not accessible until the next restart.

3. Restart the deployment with 

    `kc.sh start --optimized`

**Comment by sschu** [comment]
Its not even necessary to start with a single replica, Keycloak will lock the database while doing schema changes...

**Reply by sirishkumar** [reply]
@sschu 
Thanks for the info. And regarding the arguments, should i use --spi-connections-jpa-default-migration-strategy=update which is from [migration guide](https://www.keycloak.org/docs/latest/upgrading/index.html#migrating-to-17-0-0) or  --spi-connections-jpa-legacy-migration-strategy=update from [GitHub documentation](https://github.com/keycloak/keycloak/blob/main/docs/guides/src/main/server/db.adoc)

**Reply by sschu** [reply]
Since this is the default, you don't have to specify it at all.

**Reply by sirishkumar** [reply]
@sschu  Is it mentioned somewhere in the documentation?

**Reply by sschu** [reply]
https://www.keycloak.org/docs/latest/upgrading/index.html#_migrate_db

**Reply by sirishkumar** [reply]
@sschu  Thanks for the clarification and documentation reference. I was a bit confused because of the terms new installation and first time in the statement "By default the database is automatically migrated when you start the new installation for the first time" as it is neither a new installation and we upgrade only once.

**Comment by sirAlexander** [comment]
@sirishkumar how did the migration go, was the jump from 15 to 19 seamless,  what docs helped.. what should one look out for?

**Reply by sirishkumar** [reply]
@sirAlexander Sorry for the delay in replying. It went smooth without any issues. From my observations, there were no changes in DB schema except one additional index, so there were no issues on the db side. 

For me i referred following
- https://www.keycloak.org/server/all-config
- https://www.keycloak.org/server/containers because of changes in config options
- https://stackoverflow.com/questions/72426072/how-to-configure-keycloak-18-running-http-only-in-production-mode
- https://www.keycloak.org/server/logging
- https://www.keycloak.org/server/caching


Since every environment is different, you may face different issues. Let me know if you have any specific questions, will try to answer them

**Reply by sirAlexander** [reply]
Thanks @sirishkumar I appreciate you taking the time to respond

**Comment by sirishkumar** [comment]
One issue we observed was we have some issues on firefox, but after all cookies, it worked fine

**Comment by Msquared63** [comment]
I'm trying to upgrade the from 12.0.4 to 21.0.1 with a aws postgres db.  But the migration keeps timing out after 5 mins and I can't figure out how to increase it.  Any suggestions?

**Reply by sschu** [reply]
You can increase the transaction timeout using Quarkus properties: e.g. quarkus.transaction-manager.default-transaction-timeout=5M, see https://quarkus.io/guides/all-config#quarkus-narayana-jta_quarkus.transaction-manager.default-transaction-timeout

**Reply by Msquared63** [reply]
Thank you!  That worked.
